odoo/models.py

```diff
@@ -1637,7 +1637,22 @@ def _name_search(self, name='', args=None, operator='ilike', limit=100, name_get
         # optimize out the default criterion of ``like ''`` that matches everything
         elif not (name == '' and operator in ('like', 'ilike')):
             aggregator = expression.AND if operator in expression.NEGATIVE_TERM_OPERATORS else expression.OR
-            domain = aggregator([[(field_name, operator, name)] for field_name in search_fnames])
+            domains = []
+            for field_name in search_fnames:
+                field_name_list = field_name.split('.')
+                model = self
+                field, others = model._fields[field_name_list[0]], field_name_list[1:] if len(field_name_list) > 1 else []
+                while field.relational and len(others): # if the fname is a field on a related model (eg: partner_id.name), get the terminal field
+                    model = self.env[field.comodel_name]
+                    field, others = model._fields[others[0]], others[1:] if len(others) > 1 else []
```

Comment:
I find this more readable (I hope it's also correct because it's untested :see_no_evil: )

```suggestion
                fname, *others = field_name.split('.')
                field, model = self._fields[fname], self
                while others: # if the fname is a field on a related model (eg: partner_id.name), get the terminal field
                    fname, *others = others
                    field, model = model._fields[fname], self.env[field.comodel_name]
```

Reply:
Why shouldn't we?

The original intent of the _rec_names_search logic was introduced in https://github.com/odoo/odoo/pull/86588 to reduce the bloat of similar name_search methods doing a simple search.  Adding that assertion would go against that effort.

Reply:
It was working before in 16.0, so we shouldn't break that behavior anyway.

Reply:
Ok. BTW, your suggestion is much cleaner.

Reply:
This one is even simpler:
```suggestion
                model = self
                for fname in field_name.split('.'):
                    field = model._fields[fname]
                    model = self.env.get(field.comodel_name)
```

Reply:
If you are ok with `model` being `None` at the end :shrug:

Reply:
There is no helper yet, I though about creating one, but it is not a priority because there aren't as many use cases.